### DISCUSSION TITLE
JavaScript에서 Onclick오류 (Onclick Error in JavaScript)

### DISCUSSION THREAD

**Discussion by alsgpwns** [community/community#21477] [Other Feature Feedback, Questions, & Ideas]
<p>달력을 만들고 있습니다.<br>
I found an error while coding the calendar.</p>
<p><img src="https://user-images.githubusercontent.com/45223821/181064921-516d692a-d31a-47ed-b919-713f15481c3a.png" alt="달력" data-base62-sha1="aiwf8W1RVk1DGl23ExfJX6SxAGL" width="689" height="436"></p>
<p>왼쪽 버튼을 클릭했을 때 저번 달로 이동 되고(보여지는)<br>
오른쪽 버튼을 클릭했을 때 다음 달로 이동 되는 함수를 만들었습니다.</p>
<p>If you click the left button, it moves to the last month (as shown).<br>
If you click the right button, I create a function that moves to the next month.</p>
<pre><code class="lang-auto">
&lt;button id="PreMonth" type="button" value="click" onclick="onRightClickTest();"&gt;&gt;&lt;/button&gt;

&lt;button id="LastMonth" type="button" value="click" onclick="onLeftClickTest();"&gt;&lt;&lt;/button&gt;`;

function onLeftClickTest() {
    firstdayInst = new Date(yyyy, mm-1, 1); 
    getfirstday = firstdayInst.getDay(); 
    lastdayInst = new Date(yyyy,mm-1,0);
    lastday = lastdayInst.getDate();
    showMonth(mm-1);
    showDay();
    showDate();
}

function onRightClickTest() {

    firstdayInst = new Date(yyyy, mm+1, 1); 
    getfirstday = firstdayInst.getDay(); 
    lastdayInst = new Date(yyyy,mm+1,0);
    lastday = lastdayInst.getDate();
    showMonth(mm+1);
    showDay();
    showDate();
    
}
</code></pre>
<p>하지만 작동이 되지 않습니다ㅠㅠ<br>
그래서 함수 밖으로 (아래와 같이) 코딩하면 작동됩니다ㅠ__ㅠ<br>
왜일까요?</p>
<p>But it doesn’t work.ᅲᅲ<br>
So, if  I code out of the function (as shown below), it works.ᅲ__ᅲ<br>
Why?</p>
<pre><code class="lang-auto">firstdayInst = new Date(yyyy, mm+1, 1); 
getfirstday = firstdayInst.getDay(); 
lastdayInst = new Date(yyyy,mm+1,0);
lastday = lastdayInst.getDate();
showMonth(mm+1);
showDay();
showDate();
</code></pre>
<p>혹시 몰라 전체코드를 첨부하겠습니다.<br>
I will attach the full code just in case.<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/alsgpwns/-HTML-CSS-JS-Make-a-calendar/tree/master/%EC%BD%94%EB%93%9C%20%EC%88%98%EC%A0%95" target="_blank" rel="nofollow noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/45223821/181064923-86619b84-14bd-4c89-9f83-fcc342a028f9.jpeg" class="thumbnail onebox-avatar" width="400" height="400">

<h3><a href="https://github.com/alsgpwns/-HTML-CSS-JS-Make-a-calendar/tree/master/%EC%BD%94%EB%93%9C%20%EC%88%98%EC%A0%95" target="_blank" rel="nofollow noopener">alsgpwns/-HTML-CSS-JS-Make-a-calendar</a></h3>

<p>[HTML+CSS+JS] Make a calendar. Contribute to alsgpwns/-HTML-CSS-JS-Make-a-calendar development by creating an account on GitHub.</p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by SethClydesdale** [comment, marked as answer]
<p>Hello <a class="mention" href="/alsgpwns">@alsgpwns</a>, welcome to the GitHub Support Community!</p>
<p>Since you’re incrementing and decrementing numbers, I would recommend using the <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Decrement" rel="nofollow noopener"><code>--</code></a> and <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Increment" rel="nofollow noopener"><code>++</code></a> operators. This way the changes made to <code>mm</code> will be stored temporarily in memory and allow you to move forward or backward.</p>
<p>Here is the updated code (see the IF statements at the top of each function):</p>
<pre><code class="lang-javascript">function onLeftClickTest() {
    if (mm != 0) mm--; // decrement mm if NOT 0 (January)
    firstdayInst = new Date(yyyy, mm, 1); 
    getfirstday = firstdayInst.getDay(); 
    lastdayInst = new Date(yyyy,mm,0);
    lastday = lastdayInst.getDate();
    showMonth(mm);
    showDay();
    showDate();
    
    document.getElementById('CalendarBody').innerHTML=text+
     `
    &lt;button id="PreMonth" type="button" value="click" onclick="onRightClickTest();"&gt;&gt;&lt;/button&gt;
    &lt;button id="LastMonth" type="button" value="click" onclick="onLeftClickTest();"&gt;&lt;&lt;/button&gt;`;
}

function onRightClickTest() {
    if (mm != 11) mm++; // increment mm if NOT 11 (December)
    firstdayInst = new Date(yyyy, mm, 1); 
    getfirstday = firstdayInst.getDay(); 
    lastdayInst = new Date(yyyy,mm,0);
    lastday = lastdayInst.getDate();
    showMonth(mm);
    showDay();
    showDate();
    
    document.getElementById('CalendarBody').innerHTML=text+
     `
    &lt;button id="PreMonth" type="button" value="click" onclick="onRightClickTest();"&gt;&gt;&lt;/button&gt;
    &lt;button id="LastMonth" type="button" value="click" onclick="onLeftClickTest();"&gt;&lt;&lt;/button&gt;`;
}
</code></pre>
<p>If you want to reset the month when you get to the end, you can add an ELSE statement, like the one below.</p>
<pre><code class="lang-javascript">else {
  mm = 11 || 0; // 11 for onRightClick OR 0 for onLeftClick
}
</code></pre>
<p>This’ll also let you increment or decrement other variables, such as the year.</p>

**Comment by alsgpwns** [comment]
<p>Hi :slight_smile:<br>
Thank you so much for your comment. :grinning: :grinning: :grinning:<br>
It works well without adding ‘if’.<br>
The reason my code didn’t work is because I didn’t add a button?<br>
(Code below)</p>
<pre><code class="lang-auto">document.getElementById('CalendarBody').innerHTML=text+
     `
    &lt;button id="PreMonth" type="button" value="click" onclick="onRightClickTest();"&gt;&gt;&lt;/button&gt;
    &lt;button id="LastMonth" type="button" value="click" onclick="onLeftClickTest();"&gt;&lt;&lt;/button&gt;`;
</code></pre>

**Comment by SethClydesdale** [comment]
<p>You’re welcome.</p>
<p>Yes, initially, clicking the buttons did not update the HTML (at least with the provided code). Because of that, you couldn’t see any changes.</p>

**Comment by alsgpwns** [comment]
<p>Thank you.<br>
With your help, I was able to make the results like the picture below.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/7/73afc466ef455e8f3eb8896a031f6a23659e32bd.gif" data-download-href="/uploads/short-url/gvpui6FtTg1VOEl8PsW4RcqDGZD.gif?dl=1" title="애니메이션1" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/45223821/181064933-039c544c-512a-48db-b13f-4e1eda5f5019.gif" alt="애니메이션1" data-base62-sha1="gvpui6FtTg1VOEl8PsW4RcqDGZD" width="671" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/7/73afc466ef455e8f3eb8896a031f6a23659e32bd_2_671x500.gif, https://aws1.discourse-cdn.com/github/original/2X/7/73afc466ef455e8f3eb8896a031f6a23659e32bd.gif 1.5x, https://aws1.discourse-cdn.com/github/original/2X/7/73afc466ef455e8f3eb8896a031f6a23659e32bd.gif 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/7/73afc466ef455e8f3eb8896a031f6a23659e32bd_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">애니메이션1</span><span class="informations">693×516 483 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
